### PR TITLE
Refactor QueryExecutor (WIP)

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request moves QueryExecutor from the `tsdb` package to the `influxql` package and adds additional `StatementExecutor` interfaces for each statement type.

## TODO

- [x] `SET PASSWORD`
- [x] `SELECT INTO`
- [x] Fix empty result returns
- [x] Fix `meta` test package

---

/cc @jwilder @dgnorton @joelegasse @e-dard

### CODE REVIEW

File: cmd/influxd/run/query_executor.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import (
+	"github.com/influxdata/influxdb/cluster"
+	"github.com/influxdata/influxdb/influxql"
+	"github.com/influxdata/influxdb/monitor"
+	"github.com/influxdata/influxdb/services/meta"
+	"github.com/influxdata/influxdb/tsdb"
+)
+
+// NewQueryExecutor returns an instance of QueryExecutor with registered statement executors.
+func NewQueryExecutor(mc *meta.Client, s *tsdb.Store, m *monitor.Monitor, pw *cluster.PointsWriter) *influxql.QueryExecutor {
```
Comment (pauldix) [inline cmd/influxd/run/query_executor.go]: `PointsWriter` not needed? Was confused why this was getting passed into a query executor

File: influxql/query_executor.go [hunk 2]
```diff
@@ -0,0 +1,423 @@
...
+	ShowTagKeysStatementExecutor           ShowTagKeysStatementExecutor
+	ShowTagValuesStatementExecutor         ShowTagValuesStatementExecutor
+	ShowUsersStatementExecutor             ShowUsersStatementExecutor
+
+	// Defaults statements to use default database & retention policy
+	// if they are not specified in measurements.
+	StatementNormalizer StatementNormalizer
+
+	// Rewrites statements into other statements, as needed.
+	StatementRewriter *StatementRewriter
```
Comment (joelegasse) [inline influxql/query_executor.go]: Why is this exposed and a pointer, or even a named type? `StatementRewriter` is defined as an empty struct, and the (pointer) receiver value is never used, other than to call methods. This field should be taken out and non-exported functions should be used instead.

**Comment by benbjohnson** [comment]
This PR is ready for review.

There is one outstanding issue where the `monitor.Monitor` depends on `cluster.PointsWriter` but now that the `QueryExecutor` exists in `cluster` it creates a circular dependency. I'm trying to figure out the simplest way to resolve that.

**Comment by jsternberg** [comment]
Looks fine to me so far, but the sections that are commented out should either be fixed if they are needed or removed if they are no longer needed.

**Comment by benbjohnson** [comment]
@jsternberg I implemented a smaller wrapper type for `cluster.PointsWriter` so it could be injected into `monitor.Monitor` in 9bebc2b. :shipit:?

File: services/meta/service_test.go [hunk 3]
```diff
@@ -567,83 +559,40 @@ func TestMetaService_Subscriptions_Create(t *testing.T) {
...
+		t.Fatalf("unexpected error: %s", err)
 	}
 
-	res := c.ExecuteStatement(mustParseStatement(`SHOW SUBSCRIPTIONS`))
-	if res.Err != nil {
-		t.Fatal(res.Err)
-	} else if got, exp := len(res.Series), 1; got != exp {
-		t.Fatalf("unexpected response.\n\ngot: %d series\nexp: %d\n", got, exp)
-	}
+	// FIXME(benbjohnson): Convert to metadata reads.
```
Comment (jwilder) [inline services/meta/service_test.go]: FIXME

File: services/meta/service_test.go [hunk 4]
```diff
@@ -567,83 +559,40 @@ func TestMetaService_Subscriptions_Create(t *testing.T) {
...
-	} else if got, exp := len(res.Series), 1; got != exp {
-		t.Fatalf("got %d series, expected %d", got, exp)
-	}
-
-	exp := `{"series":[{"name":"db0","columns":["retention_policy","name","mode","destinations"],"values":[["default","sub0","ALL",["udp://example.com:9090"]]]}]}`
-	got := mustMarshalJSON(res)
-	if got != exp {
-		t.Fatalf("unexpected response.\n\ngot: %s\nexp: %s\n", got, exp)
-	}
+	// FIXME(benbjohnson): Convert to metadata reads.
```
Comment (jwilder) [inline services/meta/service_test.go]: FIXME

File: services/meta/service_test.go (hunk 3 above)
Comment (benbjohnson) [inline services/meta/service_test.go]: Weird. I swear I removed these sections. This needs to be tested at the `QueryExecutor` level so I wasn't going to rewrite them here.

**Comment by jwilder** [comment]
:+1:

**Comment by e-dard** [comment]
LGTM :+1:

**Pull request merged**